mediorum/server/serve_blob.go

```diff
@@ -156,82 +156,68 @@ func (ss *MediorumServer) serveBlob(c echo.Context) error {
 		c.SetParamValues(values...)
 	}
 
-	logger := ss.logger.With("cid", cid)
 	key := cidutil.ShardCID(cid)
 
-	// return 403 if the requested CID is delisted
-	shouldCheckDelistStatus := true
-	if checkedDelistStatus, exists := c.Get("checkedDelistStatus").(bool); exists && checkedDelistStatus {
-		shouldCheckDelistStatus = false
-	}
-	if shouldCheckDelistStatus && ss.isCidBlacklisted(ctx, cid) {
-		logger.Info("cid is blacklisted")
-		return c.String(403, "cid is blacklisted by this node")
-	}
-
 	// if the client provided a filename, set it in the header to be auto-populated in download prompt
 	filenameForDownload := c.QueryParam("filename")
 	if filenameForDownload != "" {
 		contentDisposition := mime.QEncoding.Encode("utf-8", filenameForDownload)
 		c.Response().Header().Set("Content-Disposition", fmt.Sprintf(`attachment; filename="%s"`, contentDisposition))
 	}
 
-	// check our bucket and serve the file if we have it
-	if attrs, err := ss.bucket.Attributes(ctx, key); err == nil && attrs != nil {
-		isAudioFile := strings.HasPrefix(attrs.ContentType, "audio")
-		// detect mime type and block mp3 streaming outside of the /tracks/cidstream route
-		if !strings.Contains(c.Path(), "cidstream") && isAudioFile {
-			return c.String(401, "mp3 streaming is blocked. Please use Discovery /v1/tracks/:encodedId/stream")
-		}
+	blob, err := ss.bucket.NewReader(ctx, key, nil)
 
-		blob, err := ss.bucket.NewReader(ctx, key, nil)
-		if err != nil {
-			return err
+	if err != nil {
+		// don't redirect if the client only wants to know if we have it (ie localOnly query param is true)
+		if localOnly, _ := strconv.ParseBool(c.QueryParam("localOnly")); localOnly {
+			return c.String(404, "blob not found")
 		}
-		defer blob.Close()
 
-		// v2 file listen
-		if isAudioFile {
-			go ss.logTrackListen(c)
-		} else {
-			// images: cache 30 days
-			c.Response().Header().Set(echo.HeaderCacheControl, "public, max-age=2592000, immutable")
+		// redirect to it
+		host := ss.findNodeToServeBlob(ctx, cid)
+		if host == "" {
+			return c.String(404, "blob not found")
 		}
 
-		if c.Request().Method == "HEAD" {
-			return c.NoContent(200)
-		}
+		dest := ss.replaceHost(c, host)
+		query := dest.Query()
+		query.Add("allow_unhealthy", "true") // we confirmed the node has it, so allow it to serve it even if unhealthy
+		dest.RawQuery = query.Encode()
+		return c.Redirect(302, dest.String())
+	}
 
-		if isAudioFile {
-			go ss.recordMetric(StreamTrack)
-			http.ServeContent(c.Response(), c.Request(), cid, blob.ModTime(), blob)
-			return nil
+
+	if c.Request().Method == "HEAD" {
+		return c.NoContent(200)
+	}
+
+	defer blob.Close()
```

Comment:
this line needs to be above the no content early return or we will end up with unclosed file handlers